File: leetcode/c++/209_minimum-size_subarray_sum.cpp
// https://leetcode.cn/problems/minimum-size-subarray-sum

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int minSubArrayLen(int target, vector<int>& nums) {
        int res = INT_MAX, n = nums.size(), sum = 0;
        for (int i = 0, j = 0; j < n; ++j) {
            sum += nums[j];
            while (sum >= target) {
                res = min(res, j - i + 1);
                sum -= nums[i++];
            }
        }
        return res == INT_MAX? 0 : res;
    }
};

// 不定长滑动窗口，用while就完事了，只要当前的和大于等于target，就收缩窗口
// O(n)，O(1)

File: leetcode/golang/209_minimum-size_subarray_sum.go
func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}

func minSubArrayLen(target int, nums []int) int {
    res, n, sum := math.MaxInt32, len(nums), 0
    for i, j := 0, 0; j < n; j++ {
        sum += nums[j]
        for sum >= target {
            res = min(res, j - i + 1)
            sum -= nums[i]
            i++
        }
    }
    if res == math.MaxInt32 {
        return 0
    }
    return res
}